assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -81,6 +173,78 @@ spec:
         - name: ssl-certs-host
           mountPath: /etc/ssl/certs
           readOnly: true
+      {{- if eq (int .Values.apiserver.replicas) 1 }}
+        - name: data
+          mountPath: /run/kube-apiserver
+      - name: haproxy
+        image: haproxy:2.1.3-alpine
+        volumeMounts:
+        - name: data
+          mountPath: /run/kube-apiserver
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -xe
+          export ADDRESS=$(cat /run/kube-apiserver/address)
+          # Make sure initContainer generated kube-apiserver address.
+          if [ -z $ADDRESS ]; then
+            echo "ADDRESS not found"
+            exit 1
+          fi
+          echo "Connecting to $ADDRESS:{{ template "port" . }}"
+          # We use TCP readiness probe and HAProxy does not reject connections if no backend is available,
+          # so we wait until kube-apiserver is available here, so readiness of haproxy container represents
+          # readiness of kube-apiserver, as kube-apiserver cannot have readiness probe set, as it listens
+          # on random IP address.
+          until nc -zv $ADDRESS {{ template "port" . }}; do sleep 1; done
+          echo "Connected"
+          # From https://github.com/docker-library/haproxy/blob/master/Dockerfile-debian.template#L70
+          exec haproxy -f /run/kube-apiserver/haproxy.cfg
+        readinessProbe:
+          tcpSocket:
+            port: {{ template "port" . }}
+      initContainers:
+      - name: config-generator
+        image: haproxy:2.1.3-alpine
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -xe
+          export ADDRESS="127.$(shuf -i 0-255 -n 1).$(shuf -i 0-255 -n 1).$(shuf -i 1-253 -n 1)"
+          echo $ADDRESS > /run/kube-apiserver/address
+          echo "defaults
+            # Do TLS passthrough
+            mode tcp
+            # Required values for both frontend and backend
+            timeout connect 5s
+            timeout client 30s
+            timeout client-fin 30s
+            timeout server 30s
+            timeout tunnel 21d
+
+          frontend kube-apiserver-internal
+            bind $POD_IP:{{ template "port" . }}
+            default_backend kube-apiserver
```

Comment:
Not sure I follow... this is to handle the case of a single controller node configured without HAProxy?

I thought we didn't allow those kind of installations at the lokoctl level. Do we want to support those setups or just have this "single node, then eat HAProxy" to simplify the yamls?

Reply:
I don't understand.

>Not sure I follow... this is to handle the case of a single controller node configured without HAProxy?

We don't have a case of single controller node without HAProxy. With single controller node, we deploy HAProxy to handle SO_REUSEPORT for us.

Reply:
Okay... so, I think I'm confused on why we need two different listeners on HAProxy then?

Sorry, I think I'm not getting the full picture or kind of tired :D

Reply:
>Sorry, I think I'm not getting the full picture or kind of tired :D

I just realized that there is so many moving pieces to understand for that, so I understand the confusion :smile:. I Added a comment at the top of the file explaining, why we need to listen on 2 IP addresses.

Reply:
I'm kind of lost to which comment do you refer to. But thanks! Maybe the comment is this: https://github.com/kinvolk/lokomotive/pull/32/files#diff-12be5db389a0fefe0038d10211a890b4R1-R8?

I think I got it now, more tricky than I thought originally! :-D